src/chkstat.cpp

```diff
@@ -682,7 +684,9 @@ std::string Chkstat::getPathFromProc(const FileDesc &fd) const {
 int Chkstat::processEntries() {
     size_t errors = 0;
 
-    if (m_apply_changes && !checkHaveProc()) {
+    m_proc_mount_avail = procState();
```

Comment:
There's no TOCTOU at least no new one compared to the original code. Of course /proc could disappear or appear in the meanwhile but that is not a scenario that the tool is concerned with. I don't believe it would be possible to protect against a parallel mount/unmount operation unless we split off a dedicated mount namespace and mount proc on our own.

I already thought about getting rid of the `ProcMountState` enum altogether and just keep a boolean around that gets initialized early on. This could be done in the Chkstat constructor already, thereby keeping `m_proc_mount_avail` const. The only downside to this is that the check would happen even if it is unnecessary e.g. when running something like `chkstat --print-variables`.

Reply:
Ok, perhaps worrying about TOCTOU is a bit too sci-fi. Two things could happen:
1. `proc` is not mounted when we check, and is surreptitiously mounted later:
    In this case the tool will just operate in a reduced capacity, plainly visible to the user, and will not do any damage.
2. `proc` is mounted when we check, and is surreptitiously unmounted later:
    In this case the tool will most likely just crash upon first access of a path under `proc`, also plainly visible to the user. Probably no damage is done in this case either :thinking: 

As for the `ProcMountState`, leaving only a boolean would indeed be simpler. And I don't think checking this unconditionally would add a significant overhead. Your call whether it's worth it or if leaving things as they are suffices.